fix(nodes-base): fix and harmonize all primaryDocumentation links

mieky opened the pull request:
This is a housekeeping 🏠 🔨  PR to bring all the `primaryDocumentation` links in nodes-base up to date and help @StarfallProjects keep them tidy.

With the help of a few custom scripts, I was able to locate and fix a couple of 404 links, malformed JSONs, and unnecessary redirects. The bulk of this change is to crawl all primaryDocumentation links in node descriptor files, and for those that would be redirected (HTTP 301), replace the JSON with their actual target.

Finally, ran `npx prettier --write nodes/**/*.node.json` to pretty-print the files.

With this change, all node URLs should take us directly to their correct documentation pages, and we can further clean up some of the manual redirects in the docs.

mieky committed:
fix(nodes-base): fix and harmonize all primaryDocumentation links

StarfallProjects reviewed (dismissed):
Looks good to me.

ivov reviewed (commented):
1. URL is case-insensitive so the link will work, but should we be allowing for this casing divergence? `n8n-nodes-base.activecampaigntrigger` vs. `n8n-nodes-base.activeCampaignTrigger` The canonical name is camelCased after the prefix. No immediate risk that I can think of, but I would avoid this divergence if we can help it. 
2. `With the help of a few custom scripts, I was able to locate and fix a couple of 404 links, malformed JSONs, and unnecessary redirects.` Can you add a task to automate this in future? Via CI or an n8n workflow.
3. `npx prettier --write nodes/**/*.node.json` Can you update the `format` command to format these files as well?

StarfallProjects commented:
> 1. URL is case-insensitive so the link will work, but should we be allowing for this casing divergence? `n8n-nodes-base.activecampaigntrigger` vs. `n8n-nodes-base.activeCampaignTrigger` The canonical name is camelCased after the prefix. No immediate risk that I can think of, but I would avoid this divergence if we can help it.

Our hosting automatically lowercases URLs (this is the background to making the links lowercase)

krynble reviewed (dismissed):
Nice

Perhaps we can also change https://github.com/n8n-io/n8n/blob/6ac442a2accdeb4e51a3333ac6c45c8bfde0608d/packages/editor-ui/src/components/NodeSettingsTabs.vue#L49 so it also reflects the new URLs? Without this change n8n continues sending users to the wrong URL, requiring the 301 to be there forever.

mieky committed:
feat(workflow, cli): expose documentation links to UI via node codex

mieky committed:
fix(editor-ui): link to correct node and credential documentation URLs

mieky committed:
config(nodes-base): update 'format' script to also format node descriptor json

mieky committed:
chore: fix outdated links to node reference documentation

mieky committed:
fix(cli): add missing dependency for lodash.pick

mieky commented on packages/cli/src/LoadNodesAndCredentials.ts:
```diff
@@ -428,19 +429,23 @@ class LoadNodesAndCredentialsClass {
 	getCodex(filePath: string): CodexData {
 		// eslint-disable-next-line global-require, import/no-dynamic-require, @typescript-eslint/no-var-requires
-		const { categories, subcategories, alias } = require(`${filePath}on`); // .js to .json
+		const { categories, subcategories, resources: allResources, alias } = require(`${filePath}on`); // .js to .json
+
+		const resources = pick(allResources, ['primaryDocumentation', 'credentialDocumentation']);
+
 		// eslint-disable-next-line @typescript-eslint/no-unsafe-return
 		return {
 			...(categories && { categories }),
 			...(subcategories && { subcategories }),
+			...(resources && { resources }),
 			...(alias && { alias }),
 		};
 	}
```
My thinking is that in the long run the only maintainable way to maintain the documentation links is to get them directly from the node descriptor (codex). I now exposed `resources.primaryDocumentation` and `resources.credentialDescription` so that the UI can access and use those, instead of having to generate the URLs dynamically from parts.

mieky commented on packages/editor-ui/src/components/CredentialEdit/CredentialConfig.vue:
```diff
@@ -206,9 +206,9 @@ export default mixins(restApi).extend({
+			return isCommunityNode ?
 				'' : // Don't show documentation link for community nodes if the URL is not an absolute path
-				`https://docs.n8n.io/credentials/${type.documentationUrl}/?utm_source=n8n_app&utm_medium=left_nav_menu&utm_campaign=create_new_credentials_modal`;
+				`${BUILTIN_CREDENTIALS_DOCS_URL}${type.documentationUrl}/?utm_source=n8n_app&utm_medium=left_nav_menu&utm_campaign=create_new_credentials_modal`;
```
Parametrized the credential link URL prefix.

mieky commented on packages/editor-ui/src/components/NodeSettingsTabs.vue:
```diff
@@ -45,11 +40,21 @@ export default mixins(
+			const utmTags = '?utm_source=n8n_app&utm_medium=node_settings_modal-credential_link' +
+				'&utm_campaign=' + nodeType.name;
+
+			// Built-in node documentation available via its codex entry
+			const primaryDocUrl = nodeType.codex?.resources?.primaryDocumentation?.[0]?.url;
+			if (primaryDocUrl) {
+				return primaryDocUrl + utmTags;
+			}
```
All our built-in nodes (with the exception of the StickyNote, I think) have a primary documentation link defined in their descriptor, so let's use that directly where it applies.

mieky commented on packages/workflow/src/Interfaces.ts:
```diff
@@ -1618,9 +1618,17 @@ export interface IStatusCodeMessages {
 export type CodexData = {
 	categories?: string[];
 	subcategories?: { [category: string]: string[] };
+	resources?: {
+		credentialDocumentation?: DocumentationLink[];
+		primaryDocumentation?: DocumentationLink[];
+	};
```
Exposing documentation links as part of a node's CodexData.

mieky commented:
@krynble @ivov Thanks for the great input! I refactored how the documentation links are formed, and so far all the ones I clicked would seem to open the correct URL directly. It's possible there's a case I might have missed, but the logic overall seems quite solid. I also updated some more old links that obviously would either get straight up redirected or were just broken.

Also created a N8N-4791 for including the helper scripts in the repo, to make it easier to deal with these links in the future. Pointed my most notable changes as line comments.

krynble reviewed (approved):
LGTM!

janober commented:
Got released with `n8n@0.196.0`
